### PR TITLE
INS: added automatic register checking for IMU drivers

### PR DESCRIPTION
**Pull request by tridge**
This adds register checking for IMU drivers, so that if a sensor goes bad in flight it will be marked as in error and will try to auto-fix the register value.

Depends on #5184

### CODE REVIEW

File: libraries/AP_InertialSensor/AP_InertialSensor_LSM9DS0.cpp [hunk 1]
```diff
@@ -468,6 +472,10 @@ bool AP_InertialSensor_LSM9DS0::_hardware_init()
         goto fail_whoami;
     }
 
+    // setup for register checking
+    _dev_gyro->setup_checked_registers(6);
+    _dev_accel->setup_checked_registers(5);
```
Comment (OXINARF) [inline libraries/AP_InertialSensor/AP_InertialSensor_LSM9DS0.cpp]: Have I counted wrong or is there a reason for these to have one more register than those marked as checked?

File: libraries/AP_InertialSensor/AP_InertialSensor_MPU9250.cpp [hunk 2]
```diff
@@ -518,7 +519,7 @@ bool AP_InertialSensor_MPU9250::_read_sample()
     uint8_t n_samples;
     uint16_t bytes_read;
     uint8_t *rx = _fifo_buffer;
-
+    
     if (!_block_read(MPUREG_FIFO_COUNTH, rx, 2)) {
         hal.console->printf("MPU9250: error in fifo read\n");
         return true;
```
Comment (OXINARF) [inline libraries/AP_InertialSensor/AP_InertialSensor_MPU9250.cpp]: Missed a change to `goto check_registers`?

File: libraries/AP_BoardConfig/px4_drivers.cpp [hunk 3]
```diff
@@ -766,6 +765,87 @@ void AP_BoardConfig::px4_setup_peripherals(void)
...
+/*
+  auto-detect board type
+ */
+void AP_BoardConfig::px4_autodetect(void)
+{
+    if (px4.board_type == PX4_BOARD_TEST_V1 ||
+        px4.board_type == PX4_BOARD_TEST_V2 ||
+        px4.board_type == PX4_BOARD_TEST_V3) {
+        // these were old test values
+        px4.board_type.set(0);
```
Comment (OXINARF) [inline libraries/AP_BoardConfig/px4_drivers.cpp]: Do we really need this? It has been in master for less than 48 hours.

File: libraries/AP_BoardConfig/px4_drivers.cpp [hunk 4]
```diff
@@ -766,6 +765,87 @@ void AP_BoardConfig::px4_setup_peripherals(void)
...
+        hal.console->printf("Detected PixhawkMini\n");
+    } else if (spi_check_register(HAL_INS_LSM9DS0_A_NAME, LSMREG_WHOAMI, LSM_WHOAMI_LSM303D) &&
+               (spi_check_register(HAL_INS_MPU60x0_NAME, MPUREG_WHOAMI, MPU_WHOAMI_MPU60X0) ||
+                spi_check_register(HAL_INS_ICM20608_NAME, MPUREG_WHOAMI, MPU_WHOAMI_ICM20608) ||
+                spi_check_register(HAL_INS_MPU9250_NAME, MPUREG_WHOAMI, MPU_WHOAMI_MPU9250))) {
+        // classic or upgraded Pixhawk1
+        px4.board_type.set(PX4_BOARD_PIXHAWK);
+        hal.console->printf("Detected Pixhawk\n");
+    } else {
+        px4_sensor_error("Unable to detect board type");
```
Comment (OXINARF) [inline libraries/AP_BoardConfig/px4_drivers.cpp]: What about PH2SLIM? There is no way for auto-detect?

**Comment by hiro2233** [comment]
I pushed a PR to tridge:pr-register-checking branch with some improvements for FIFO operation.
https://github.com/tridge/ardupilot/pull/7

Thanks.

File: libraries/AP_InertialSensor/AP_InertialSensor_LSM9DS0.cpp (hunk 1 above)
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor_LSM9DS0.cpp]: you've counted right, fixed, thanks!

File: libraries/AP_InertialSensor/AP_InertialSensor_MPU9250.cpp (hunk 2 above)
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor_MPU9250.cpp]: fixed, thanks!

File: libraries/AP_BoardConfig/px4_drivers.cpp (hunk 3 above)
Comment (tridge) [inline libraries/AP_BoardConfig/px4_drivers.cpp]: I'll probably remove them in the future, its really just for the convenience of the few people testing now

File: libraries/AP_BoardConfig/px4_drivers.cpp (hunk 4 above)
Comment (tridge) [inline libraries/AP_BoardConfig/px4_drivers.cpp]: unfortunately we can't tell the difference between a PH2SLIM and a PH2 that has a bad connector to the sensor sub-board

**Comment by tridge** [comment]
applied - thanks to everyone for the feedback!

**Pull request closed without merging**